smart_open/gcs.py

```diff
@@ -0,0 +1,524 @@
+
+        #
+        # Return unused data first
+        #
+        if len(self._buffer) >= size:
+            return self._read_from_buffer(size)
+
+        #
+        # If the stream is finished, return what we have.
+        #
+        if self._eof:
+            return self._read_from_buffer()
+
+        #
+        # Fill our buffer to the required size.
+        #
+        # logger.debug('filling %r byte-long buffer up to %r bytes', len(self._buffer), size)
+        self._fill_buffer(size)
+        return self._read_from_buffer(size)
+
+    def read1(self, size=-1):
+        """This is the same as read()."""
+        return self.read(size=size)
+
+    def readinto(self, b):
+        """Read up to len(b) bytes into b, and return the number of bytes
+        read."""
+        data = self.read(len(b))
+        if not data:
+            return 0
+        b[:len(data)] = data
+        return len(data)
+
+    def readline(self, limit=-1):
+        """Read up to and including the next newline.  Returns the bytes read."""
+        if limit != -1:
+            raise NotImplementedError('limits other than -1 not implemented yet')
+        the_line = io.BytesIO()
+        while not (self._eof and len(self._buffer) == 0):
+            #
+            # In the worst case, we're reading the unread part of self._buffer
+            # twice here, once in the if condition and once when calling index.
+            #
+            # This is sub-optimal, but better than the alternative: wrapping
+            # .index in a try..except, because that is slower.
+            #
+            remaining_buffer = self._buffer.peek()
+            if self._line_terminator in remaining_buffer:
+                next_newline = remaining_buffer.index(self._line_terminator)
+                the_line.write(self._read_from_buffer(next_newline + 1))
+                break
+            else:
+                the_line.write(self._read_from_buffer())
+                self._fill_buffer()
+        return the_line.getvalue()
+
+    def terminate(self):
+        """Do nothing."""
+        pass
+
+    #
+    # Internal methods.
+    #
+    def _read_from_buffer(self, size=-1):
+        """Remove at most size bytes from our buffer and return them."""
+        # logger.debug('reading %r bytes from %r byte-long buffer', size, len(self._buffer))
+        size = size if size >= 0 else len(self._buffer)
+        part = self._buffer.read(size)
+        self._current_pos += len(part)
+        # logger.debug('part: %r', part)
+        return part
+
+    def _fill_buffer(self, size=-1):
+        size = size if size >= 0 else self._buffer._chunk_size
+        while len(self._buffer) < size and not self._eof:
+            bytes_read = self._buffer.fill(self._raw_reader)
+            if bytes_read == 0:
+                logger.debug('reached EOF while filling buffer')
+                self._eof = True
+
+
+class SeekableBufferedInputBase(BufferedInputBase):
```

Comment:
Why do you need two separate classes?

1. BufferedInputBase
2. SeekableBufferedInputBase

Can't you get away with just one of them?

Reply:
Another great point, BufferedInputBase has been removed